### PR TITLE
DIETClassifier adds a confidence value to entity predictions

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- Update the CRF to return scores that can be used as confidence values.
- `DIETClassifier` now adds a confidence value to entity predictions.

closes https://github.com/RasaHQ/rasa/issues/5481

(Will also make a PR to tensorflow addons directly. If they take the changes and adapt the CRF, we can remove our custom CRF again.)

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
Tensorflow Addons PR: https://github.com/tensorflow/addons/pull/1935

File: rasa/nlu/classifiers/diet_classifier.py [hunk 1]
```diff
@@ -811,20 +814,24 @@ def _predict_entities(
...
+        confidence_values = {}
 
         for tag_spec in self._entity_tag_specs:
             predictions = predict_out[f"e_{tag_spec.tag_name}_ids"].numpy()
+            confidences = predict_out[f"e_{tag_spec.tag_name}_scores"].numpy()
+            confidences = [float(c) for c in confidences[0]]
             tags = [tag_spec.ids_to_tags[p] for p in predictions[0]]
 
             if self.component_config[BILOU_FLAG]:
                 tags = bilou_utils.ensure_consistent_bilou_tagging(tags)
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/diet_classifier.py:828]: can change tags/merge tags? does it affect confidences?

File: rasa/utils/tensorflow/crf.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+          output: A [batch_size, num_tags * 2] matrix of backpointers and scores.
+          new_state: A [batch_size, num_tags] matrix of new score values.
+        """
+        state = tf.expand_dims(state[0], 2)
+        transition_scores = state + self._transition_params
+        new_state = inputs + tf.reduce_max(transition_scores, [1])
+        backpointers = tf.argmax(transition_scores, 1)
+        backpointers = tf.cast(backpointers, tf.float32)
+        scores = tf.reduce_max(tf.nn.softmax(transition_scores, axis=1), [1])
+        return tf.concat([backpointers, scores], axis=1), new_state
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:61]: please add a comment why concatenate

File: rasa/utils/tensorflow/crf.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        Returns:
+          output: A [batch_size, num_tags * 2] matrix of backpointers and scores.
+          new_state: A [batch_size, num_tags] matrix of new score values.
+        """
+        state = tf.expand_dims(state[0], 2)
+        transition_scores = state + self._transition_params
+        new_state = inputs + tf.reduce_max(transition_scores, [1])
+        backpointers = tf.argmax(transition_scores, 1)
+        backpointers = tf.cast(backpointers, tf.float32)
+        scores = tf.reduce_max(tf.nn.softmax(transition_scores, axis=1), [1])
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:60]: please add a comment to explain why softmax here

File: rasa/utils/tensorflow/crf.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    mask = tf.sequence_mask(sequence_lengths, tf.shape(inputs)[1])
+    crf_fwd_cell = CrfDecodeForwardRnnCell(transition_params)
+    crf_fwd_layer = tf.keras.layers.RNN(
+        crf_fwd_cell, return_sequences=True, return_state=True
+    )
+    return crf_fwd_layer(inputs, state, mask=mask)
+
+
+def crf_decode_backward(
+    inputs: TensorLike, scores: TensorLike, state: TensorLike
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:93]: let's rename to `backpointers`
```suggestion
    backpointers: TensorLike, scores: TensorLike, state: TensorLike
```

File: rasa/utils/tensorflow/crf.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    inputs = tf.transpose(inputs, [1, 0, 2])
+    scores = tf.transpose(scores, [1, 0, 2])
+
+    def _scan_fn(state, inputs):
+        state = tf.cast(tf.squeeze(state, axis=[1]), dtype=tf.int32)
+        idxs = tf.stack([tf.range(tf.shape(inputs)[0]), state], axis=1)
+        new_tags = tf.expand_dims(tf.gather_nd(inputs, idxs), axis=-1)
+        return new_tags
+
+    output_tags = tf.scan(_scan_fn, inputs, state)
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:116]: this doesn't modify state?

File: rasa/utils/tensorflow/crf.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    state = tf.cast(state, dtype=tf.float32)
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:117]: why do we need this?

File: rasa/utils/tensorflow/crf.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+        sequence_length_less_one = tf.maximum(
+            tf.constant(0, dtype=tf.int32), sequence_length - 1
+        )
+
+        output, last_score = crf_decode_forward(
+            inputs, initial_state, transition_params, sequence_length_less_one
+        )
+
+        backpointers, scores = tf.split(output, 2, axis=2)
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:165]: please add a comment to explain why axis=2

File: rasa/utils/tensorflow/crf.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+        decode_tags = tf.squeeze(decode_tags, axis=[2])
+        decode_tags = tf.concat([initial_state, decode_tags], axis=1)
+        decode_tags = tf.reverse_sequence(decode_tags, sequence_length, seq_axis=1)
+
+        decode_scores = tf.squeeze(decode_scores, axis=[2])
+        decode_scores = tf.concat([initial_score, decode_scores], axis=1)
+        decode_scores = tf.reverse_sequence(decode_scores, sequence_length, seq_axis=1)
+
+        return decode_tags, decode_scores
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:192]: since you changed `return` please add a comment how to find the old `best_score`

File: rasa/utils/tensorflow/crf.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    sequence_length = tf.cast(sequence_length, dtype=tf.int32)
+
+    # If max_seq_len is 1, we skip the algorithm and simply return the
+    # argmax tag and the max activation.
+    def _single_seq_fn():
+        decode_tags = tf.cast(tf.argmax(potentials, axis=2), dtype=tf.int32)
+        best_score = tf.reshape(
+            tf.reduce_max(tf.nn.softmax(potentials, axis=2), axis=2), shape=[-1]
+        )
+        return decode_tags, best_score
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/crf.py:149]: is the shape of the `best_score` correct?

File: rasa/nlu/classifiers/diet_classifier.py (hunk 1 above)
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py:828]: Actually yes, the tags can change. Confidences are not considered so far. They would stay the same. Same logic is applied in `CRFEntityExtractor`. Would create a separate issue for that.

**Comment by Ghostvv** [comment]
@tabergma could you please apply my comments to tf addons PR as well

File: rasa/utils/tensorflow/crf.py (hunk 5 above)
Comment (tabergma) [reply rasa/utils/tensorflow/crf.py]: no, its not modified

File: rasa/utils/tensorflow/crf.py (hunk 8 above)
Comment (tabergma) [reply rasa/utils/tensorflow/crf.py]: added `best_score` back to the return statement

File: rasa/nlu/classifiers/diet_classifier.py (hunk 1 above)
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py:828]: https://github.com/RasaHQ/rasa/issues/6039

**Comment by vish0701** [comment]
This is a good feature to add. However, current rasa documentation at https://rasa.com/docs/rasa/nlu/entity-extraction/#entity-extraction-custom-features says 
image

Does the documentation need to be updated to indicate that now DIET returns a confidence value?

**Comment by tabergma** [comment]
@vish0701 The feature is not yet released. It is merged into master and we also updated the documentation. Once we do the next release from master, you will have the confidence value for entities coming from the DIETClassifier and the documentation will be updated as well.

**Comment by aminekechaou** [comment]
Was this already included in a release? From my understanding it is not in 1.10.14.

**Comment by tabergma** [comment]
This will be included in the Rasa 2.0 release, which should happen any day now.
